### ISSUE TITLE
Visual edit flag

### ISSUE THREAD

**Issue by quicknir** [vrapper/vrapper#639]
Visual edit can be activated in regular vim via :set ve+=block. This greatly improves the behavior of visual block mode by allowing it to select space on the screen where there are no characters. This is a pretty huge win; e.g. to add this text (with "added" aligned):

hello             added
cat_in_hat    added
a                  added

This can't be done easily without visual edit.

**Comment by albertdev** [comment]
@quicknir 
I have replaced all mentions of "visual edit" with "virtual edit" as that is what Vim calls it. I also chose to reformat your example as a code block and changed the number of spaces to work with a fixed-width font, please check that I didn't make a mistake.

At first I thought this was a duplicate of #349, but then I realized that it was about the `all` setting. While I think that Eclipse's block mode can work this way, Vrapper's needs a complete overhaul if we want to have any chance to get this working. I've marked it as a feature request, but I don't know when anyone gets to it.

**Comment by quicknir** [comment]
@albertdev Any updates on this? I know that dev on vrapper has re-invigorated a bit, thought I'd pop this up from the pile.

**Comment by albertdev** [comment]
I've been thinking some more about this and I'm seriously considering outright closing it. Eclipse support for virtual editing is still non-existent and modifying Vrapper to work with virtual caret offsets is a *huge* undertaking which I don't see happening.

The best idea I have right now is to improve support for mouse block selects, let you drag behind the line and then use Vrapper's block mode code to fill in extra spaces.

**Comment by quicknir** [comment]
You're right, that can't be done in Eclipse either. However, suppose you have this:

    hello
    goodbye
    hello

Suppose I want to delete from the 5th character onwards on all three lines. I can't really do this in vrapper. I put the cursor over the 'o' in the first hello, put it in block visual mode, go down twice, and then hit $ to go to the end of the line. It will not go to below the 'e' in goodbye. This does work in Eclipse's block selection mode, so it should (?) somehow be possible without doing something drastic.

**Comment by quicknir** [comment]
To make it even more fun, I can actually click with the mouse (I can see that vrapper's block visual uses Eclipse's block selection because the cursor changes to a crosshair) in the above example to make it work! If I put the cursor on the first 'o' and hit ctrl V, and then hold shift, I can click just below the 'e' in goodbye and get the selection I want (although now the cursor and the block selection are separated, a bit weird). So I suppose what's actually broken is the behavior of $; in this situation $ should behave specially when on the third line and take you to under the e. In vim, it actually doesn't move the cursor to under the e when you do this example; instead it leaves the cursor where it is but extends the highlighted area only on the middle line so that you can do the desired delete operation. Probably that behavior won't be possible to duplicate in vrapper.

Anyhow, hope this update was useful. If you think that it's possible to change things so that my second example is doable, great. If you think it's not practical, feel free to close.

**Comment by albertdev** [comment]
I fixed the UI part of `$` a while ago, it should show a proper selection for all text after the `e` on the second line. The problem then is that some of the internal code doing the changes will then spaz out and somehow only edits a block two characters wide. I seriously dislike the current state of the visual block code but I haven't figured out a proper way to replace it yet.

That's just "to end of line" behaviour though, it works whether or not `virtualedit` is enabled. It's still do-able because the start and end will still be somewhere in the text characters. Virtual edit on the other hand allows you to move the rectangle entirely after the text, putting the caret in a spot which you could never reach without `virtualedit`. It's that which I'm struggling to do; while I might be able to paint the caret somewhere else in the editor, keeping track of it gets hard when the usual coordinates (byte offset in the file) become totally irrelevant.

The advantage of using the mouse is that you can manipulate the visual block selection directly, then use screen coordinates to cut things up into spans of text. You don't have to keep track of byte offsets, it's only screen offsets which then get translated back to columns and line offsets. Changing Vrapper to completely work in line and column offsets is too big a challenge for me.

Anyway, I'll sleep on it.

**Comment by quicknir** [comment]
When you say you fixed the UI party, what exactly do you mean? For me it does not work; I cannot easily delete the 'o's from hello and 'bye' from goodbye simultaneously with a single visual block selection.

**Comment by albertdev** [comment]
I meant that the to-end-of-line-behaviour is supposed to be implemented. I think this looks like it's supposed to be:
![vis_block_toeol](https://cloud.githubusercontent.com/assets/1451362/13456928/b6183402-e065-11e5-9c65-c2be535ba95b.PNG)

It's just that it doesn't work 100% correctly when it should, I've documented that in some other issue.

**Comment by quicknir** [comment]
Ah ok. I can close this then; this is the behavior I really wanted; the virtual edit stuff is less important and clearly it's going to be a pain. For me the end of line never works the way you described; is there an issue where I can follow that instead?

**Comment by albertdev** [comment]
Maybe keep this one around, I might rename it and keep it open to track my current ideas.

The ones you want to watch are #663, the problem I just described is tracked as #694.
